### ISSUE TITLE
rtk query: `isSuccess` never becomes `true` and `status` is reset to `uninitialized ` upon first mount

### ISSUE THREAD

**Issue by Morriz** [reduxjs/redux-toolkit#2167]
Hi, I notice that upon first rendering of a react component with the following function body, the very first time `isSuccess` never returns with `true` after a successful request (all 3 crud methods suffer from the same):

```
export default function ({
  match: {
    params: { teamId },
  },
}: RouteComponentProps<Params>): React.ReactElement {
  useAuthzSession(teamId)
  const [formData, setFormData] = useState()
  const [deleteId, setDeleteId]: any = useState()
  const [create, { isLoading: isLoadingCreate, isSuccess: okCreate, status: statusCreate }] = useCreateTeamMutation()
  const [update, { isLoading: isLoadingUpdate, isSuccess: okUpdate, status: statusUpdate }] = useEditTeamMutation()
  const [del, { isLoading: isLoadingDelete, isSuccess: okDelete, status: statusDelete }] = useDeleteTeamMutation()
  const { data, isLoading, error } = useGetTeamQuery(
    { teamId },
    { skip: !teamId || isLoadingCreate || isLoadingUpdate || isLoadingDelete || okCreate || okUpdate || okDelete },
  )
  useEffect(() => {
    if (formData) {
      setFormData(undefined)
      if (teamId) update({ teamId, body: omit(formData, ['id']) as typeof formData })
      else create({ body: formData })
    } else if (deleteId) {
      setDeleteId()
      del({ teamId: deleteId })
    }
  }, [formData, deleteId])
  // END HOOKS
  // console.log('statusCreate: ', statusCreate)
  // console.log('isLoadingCreate: ', isLoadingCreate)
  // console.log('okCreate: ', okCreate)
  // console.log('statusUpdate: ', statusUpdate)
  // console.log('isLoadingUpdate: ', isLoadingUpdate)
  // console.log('okUpdate: ', okUpdate)
  console.log('statusDelete: ', statusDelete)
  console.log('isLoadingDelete: ', isLoadingDelete)
  console.log('okDelete: ', okDelete)
  if (okDelete || okCreate || okUpdate) return <Redirect to='/teams' />
  const team = formData || data
  const comp = !(isLoading || error) && <Team team={team} onSubmit={setFormData} onDelete={setDeleteId} />
  return <PaperLayout loading={isLoading} comp={comp} />
}
```

Running it subsequently all works well. Seems like a bug.

Log output first run (hard reload first of browser window), by providing `deleteId`:
```
statusDelete:  uninitialized
isLoadingDelete:  false
okDelete:  false
statusDelete:  pending
isLoadingDelete:  true
okDelete:  false
statusDelete:  pending
isLoadingDelete:  true
okDelete:  false
[Violation] Forced reflow while executing JavaScript took 40ms
statusDelete:  uninitialized
isLoadingDelete:  false
okDelete:  false
statusDelete:  uninitialized
isLoadingDelete:  false
okDelete:  false
[Violation] Forced reflow while executing JavaScript took 31ms
```
Second run:
```
statusDelete:  uninitialized
isLoadingDelete:  false
okDelete:  false
statusDelete:  pending
isLoadingDelete:  true
okDelete:  false
statusDelete:  pending
isLoadingDelete:  true
okDelete:  false
statusDelete:  fulfilled
isLoadingDelete:  false
okDelete:  true
```

Version used: 
```
"@reduxjs/toolkit": "^1.8.0",
"react-redux": "^7.2.6",
```

**Comment by markerikson** [comment]
I'm not sure I understand what the example code is trying to show, or what the sequence is here.

Could you put together a CodeSandbox that specifically shows this happening, and possibly also [a Replay](https://docs.replay.io/docs/recording-bug-reports-80c37d7d6753485f81497570625d06f0) with some comments pointing to what you're expecting to see and when?

**Comment by Morriz** [comment]
Hmmm. The thing is I don't know how to set up this boilerplate as I used the generator. But if you can point to a codesandbox that is preset with simple stuff like this, please point me to it so I can copy that and modify it.

**Comment by Morriz** [comment]
Could I just record a replay instead?

**Comment by phryneas** [comment]
You seem to be running the hooks' trigger functions outside of a `useEffect`, just in the component body.

That means, on every render, either `update` or `create` will be triggered - as a consequence, the component rerenders to track the new mutation that was just triggered - and triggers another mutation. You just have an endless loop, and since always the latest mutation is being watched, it will never go beyond `loading`.

And if we assume you just skipped an event handler here: since every mutation is tracked individually, you will never see the result of a mutation that was triggered in another component. Every component will start out as `uninitialized`.

Imagine having three "send money" buttons - you would not want all of them to show "sending money" and "success" if you click one of them.

**Comment by markerikson** [comment]
And that's why I was wanting to see a CodeSandbox. I wasn't sure if you were skipping showing the "run these in a click handler" step, or actually running them _while rendering_ (which is bad).

**Comment by Morriz** [comment]
Got it, I will place the trigger functions in useContext with the right deps...tnx for the quick replies!

**Comment by Morriz** [comment]
Doesn't change a thing unfortunately. Setting up a codesandbox that can handle mutations is also not straightforward it seems. Any pointers?

**Comment by Morriz** [comment]
I've update the code above to reflect the changes, and it would seem simple to analyze for you guys, as it all happens within that piece of code. Like I said, only the very first time isSuccess does not return, and there are no console messages in between. The last status update just never seems to arrive.

**Comment by Morriz** [comment]
I do see a reflow happening in between upon further inspection. Updated above as well.

**Comment by markerikson** [comment]
Honestly, we're not going to be able to solve anything here unless we can actually run the code ourselves, or see exactly what happened in a Replay.  The information in that comment simply isn't sufficient.

**Comment by Morriz** [comment]
Can I then just share the replay? codesandbox is giving me a large headache

**Comment by markerikson** [comment]
A Replay would be very helpful, yes.

**Comment by Morriz** [comment]
Ok, here are the two replays:
1. https://app.replay.io/recording/otomi-first-delete--58da31b7-98c1-4e5b-9f3e-6e895f732740

2. https://app.replay.io/recording/otomi-second-run--686cf05a-d0f7-4e79-a69a-92b49174f671

**Comment by Morriz** [comment]
Newver mind that the second run is trying to do the last fetch (and 404ing on it): it happened only after the `isSuccess: true` coming back from the delete action.

**Comment by Morriz** [comment]
I found the quirk: I created middleware that listens to mutations and sets a global dirty flag which caused in between re-renders. But I assumed that would be handled without problems. I just migrated from an openapi-axios setup that was not backed by a global store such as redux, and would never complain about these things. I am not really happy with the finicky nature of redux now that I think of it. It has to do with the fact that redux is decoupled from react and its reflows I think. It would be great if redux could deal with reflows in between state changes. Or can it? What am I overlooking?

**Comment by Morriz** [comment]
Maybe I should not set it directly but attach it to a listener in listenerMiddleware and then redux would shuffle it in the right position to handle?

**Comment by Morriz** [comment]
It would be great if I can add something to a queue of sorts so that redux finishes in flight stuff and its subscriber handling, and only then executes it

**Comment by Morriz** [comment]
I will set a timeout of a second until I find out how...pffff

**Comment by markerikson** [comment]
Honestly, I'm pretty lost at what you're trying to describe at this point :)  Again... you're talking about a bunch of things _in your app's code_, which we can't see, and so it's really hard to follow anything you're describing atm.

**Comment by Morriz** [comment]
Yeah, I'm sorry about that. It just is too much cognitive overload for me atm to transfer my thoughts and observations.
But I learned a lot, and know for a fact that I should not use a top level useState hook from middleware that triggers the whole tree rerender.

I think I just need to dispatch a custom IS_DIRTY thunk and pull it from the store in my component. It's been 2 years since I used redux (I built many apps with it over the years but it progressed over the last 2 years a lot I see).

Thanks a lot for trying to support me ;)

**Comment by phryneas** [comment]
No, you should not be using any kind of state setter at all here.

Instead of 
```
const [formData, setFormData] = useState()
const [trigger, ...] = useMutation()

useEffect(() => {trigger(formData)}, [setFormData])
<Team team={team} onSubmit={setFormData} onDelete={setDeleteId} />
```

You should be doing

```js
<Team team={team} onSubmit={
  formData => {
     if (teamId) update({ teamId, body: omit(formData, ['id']) as typeof formData })
      else create({ body: formData })
  }
} onDelete={...} />
```
or

```js
const onSubmit = useCallback(formData => {
     if (teamId) update({ teamId, body: omit(formData, ['id']) as typeof formData })
      else create({ body: formData })
  }, [teamId])

<Team team={team} onSubmit={onSubmit} onDelete={...} />
```

**Comment by phryneas** [comment]
> I found the quirk: I created middleware that listens to mutations and sets a global dirty flag which caused in between re-renders. But I assumed that would be handled without problems. I just migrated from an openapi-axios setup that was not backed by a global store such as redux, and it would never complain about these things. I am not really happy with the finicky nature of redux now that I think of it. It has to do with the fact that redux is decoupled from react and its reflows I think. It would be great if redux could deal with reflows in between state changes. Or can it? What am I overlooking?

The problem here is not with Redux. The problem here is that you bound things that should be bound to something like "a user clicks a button" into the render lifecycle of React instead on a button event. 

So you are heavily relying on React rerendering at very certain points - and that's just never a given. React renders when React wants and going forward that behaviour can and will always change. If you want to trigger a http request, do it in a click handler, not in a flimsy chain of setState - useEffect, which relies on the right things rendering in the right order (and which could trigger the useEffect multiple times when you don't want it to be triggered), over multiple components.

**Comment by Morriz** [comment]
@markerikson @phryneas I am searching for a way to subscribe to all redux errors from a central error component. Is that possible?

**Comment by phryneas** [comment]
In a Redux slice, use a `extraReducers.addMatcher` with [`isRejected`](https://redux-toolkit.js.org/api/matching-utilities#isrejected)

**Comment by Morriz** [comment]
> No, you should not be using any kind of state setter at all here.
> 
> Instead of
> 
> ```
> const [formData, setFormData] = useState()
> const [trigger, ...] = useMutation()
> 
> useEffect(() => {trigger(formData)}, [setFormData])
> <Team team={team} onSubmit={setFormData} onDelete={setDeleteId} />
> ```
> 
> You should be doing
> 
> ```js
> <Team team={team} onSubmit={
>   formData => {
>      if (teamId) update({ teamId, body: omit(formData, ['id']) as typeof formData })
>       else create({ body: formData })
>   }
> } onDelete={...} />
> ```
> 
> or
> 
> ```js
> const onSubmit = useCallback(formData => {
>      if (teamId) update({ teamId, body: omit(formData, ['id']) as typeof formData })
>       else create({ body: formData })
>   }, [teamId])
> 
> <Team team={team} onSubmit={onSubmit} onDelete={...} />
> ```
> 
> Also, if by "whole tree render" you imply that this is some kind of "top component": Put this into the lowermost component where all the children need access to it. There is no need at all to place Redux calls at the top of your component tree, they can be as far down as leaves, that's perfectly fine.

thanks so much for the clarification! I hope after refactoring this, that cache invalidation will work correctly as well.

**Comment by Nwachukwupgate** [comment]
whoa! what a day.

**Comment by Berci-Ionut** [comment]
I also have the same problem, except I am calling the trigger onClick. I also searched for an example online in order to ease the reproduction of the issue and I saw that the same behavior is happening there too (I just added some logs in order to highlight the issue when adding a new post in [this example](https://codesandbox.io/s/vibrant-tamas-mu2vn6?file=/src/features/posts/PostsManager.tsx) ). 
Here is a screenshot to highlight the issue (inside `/src/features/posts/PostsManager.tsx` ) :
![image](https://user-images.githubusercontent.com/33658866/190631197-3f0508a1-c453-478c-a701-9599f2bb678d.png)

Am I missing something?
Thank you in advice!
